### PR TITLE
implement new deployment style

### PR DESCRIPTION
**Pull request by ilgooz**

### CODE REVIEW

File: sdk/service/service.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		container: c,
+		db:        db,
+		execDB:    execDB,
+	}
+}
+
+// Create creates a new service from definition.
+func (s *ServiceSDK) Create(srv *service.Service) error {
+	// download and untar service context into path.
+	path, err := ioutil.TempDir("", "mesg-"+uuid.NewV4().String())
```
Comment (NicolasMahe) [inline sdk/service/service.go]: `ioutil.TempDir` already create an unique folder. `uuid.NewV4().String()` is not necessary.

File: sdk/service/service.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		return err
+	}
+	defer os.RemoveAll(path)
+
+	resp, err := http.Get("http://ipfs.app.mesg.com:8080/ipfs/" + srv.Source)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 200 {
+		return errors.New("service's ource code is not reachable")
```
Comment (NicolasMahe) [inline sdk/service/service.go]:
```suggestion
		return errors.New("service's source code is not reachable")
```

File: server/grpc/newcore/core.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package newcore
```
Comment (NicolasMahe) [inline server/grpc/newcore/core.go]: Can you move this file to `server/grpc/api/server.go` and the create function in `server/grpc/api/service.go` and renamed it `CreateService`

File: server/grpc/newcore/proto.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
+package newcore
```
Comment (NicolasMahe) [inline server/grpc/newcore/proto.go]: Can you move the content of this file in `server/grpc/api/service.go`

File: protobuf/service/service.proto [hunk 5]
```diff
@@ -0,0 +1,19 @@
+syntax = "proto3";
```
Comment (antho1404) [inline protobuf/service/service.proto]: maybe we can move this file in `protobuf/api/service.proto` (https://forum.mesg.com/t/protobuf-folder-structure/297)

File: sdk/service/service.go (hunk 2 above)
Comment (antho1404) [inline sdk/service/service.go]:
```suggestion
		return errors.New("service's source code is not reachable")
```

File: sdk/service/service.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/mesg-foundation/core/database"
+	"github.com/mesg-foundation/core/service"
+	"github.com/mesg-foundation/core/service/manager"
+	"github.com/mesg-foundation/core/utils/dirhash"
+	"github.com/mr-tron/base58"
+	uuid "github.com/satori/go.uuid"
+)
+
+// ServiceSDK exposes service APIs of MESG.
+type ServiceSDK struct {
```
Comment (NicolasMahe) [inline sdk/service/service.go]: What about just naming it `Service`? the filepath already give the context of sdk.

Comment (ilgooz) [reply sdk/service/service.go]: or `SDK`?

Comment (NicolasMahe) [reply sdk/service/service.go]: No, it's only about the service

File: protobuf/definition/service.proto [hunk 7]
```diff
@@ -14,6 +14,7 @@ message Service {
   repeated Event events = 6;                // The list of events this service can emit.
   repeated Dependency dependencies = 7;     // The container dependencies this service requires.
   string repository = 9;                    // Service's repository url.
+  string source = 13;                       // The hash id of service's source code on IPFS.
```
[the remark this answers is not in the document]
Comment (NicolasMahe) [reply protobuf/definition/service.proto]: After confirmation with @antho1404, we will go with only 1 field containing directly the IPFS hash 👍  closing this comment.

File: protobuf/service/service.proto (hunk 5 above)
Comment (ilgooz) [reply protobuf/service/service.proto]: done

File: sdk/service/service.go (hunk 6 above)
Comment (ilgooz) [reply sdk/service/service.go]: done

File: server/grpc/newcore/proto.go (hunk 4 above)
Comment (ilgooz) [reply server/grpc/newcore/proto.go]: done

File: server/grpc/newcore/core.go (hunk 3 above)
Comment (ilgooz) [reply server/grpc/newcore/core.go]: done

File: sdk/service/service.go (hunk 2 above)
Comment (ilgooz) [reply sdk/service/service.go]: done

File: protobuf/service/service.proto [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import "protobuf/definition/service.proto";
+
+package service;
+option go_package = "github.com/mesg-foundation/core/protobuf/service";
+
+service Service {
+  rpc Create (CreateRequest) returns (CreateResponse) {}
+}
+
+message CreateRequest {
```
[the remark this answers is not in the document]
Comment (ilgooz) [reply protobuf/service/service.proto]: done

File: protobuf/service/service.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (ilgooz) [reply protobuf/service/service.proto]: done

File: sdk/service/service.go (hunk 1 above)
Comment (ilgooz) [reply sdk/service/service.go]: done

Comment (ilgooz) [reply sdk/service/service.go]: done

File: core/main.go [hunk 9]
```diff
@@ -23,6 +24,7 @@ type dependencies struct {
 	executionDB database.ExecutionDB
 	container   container.Container
 	sdk         *sdk.SDK
+	serviceSDK  *servicesdk.ServiceSDK
```
[the remark this answers is not in the document]
Comment (ilgooz) [reply core/main.go]: done

File: protobuf/api/service.proto [hunk 10]
```diff
@@ -0,0 +1,18 @@
+syntax = "proto3";
+
+import "protobuf/definition/service.proto";
+
+package api;
+
+service ServiceX {
```
Comment (ilgooz) [inline protobuf/api/service.proto]: I put X at the end for now otherwise names conflicts for _api.Service_

File: sdk/sdk.go [hunk 11]
```diff
@@ -17,6 +18,8 @@ import (
 
 // SDK exposes all functionalities of MESG core.
 type SDK struct {
+	ServiceSDK *servicesdk.Service
```
Comment (antho1404) [inline sdk/sdk.go]: rename in `Service` we are already in the `SDK` so we can access with `sdk.Service`

File: server/grpc/api/service.go [hunk 12]
```diff
@@ -1,53 +1,53 @@
-package newcore
+package api
 
 import (
 	"github.com/mesg-foundation/core/protobuf/definition"
 	"github.com/mesg-foundation/core/service"
 )
 
-func fromProtoService(s *definition.Service) *service.Service {
+func FromProtoService(s *definition.Service) *service.Service {
```
Comment (antho1404) [inline server/grpc/api/service.go]: let's expose only this one, `fromProtoTasks`, `fromProtoEvents`... don't need to be public, they are internal functions only

File: server/grpc/server.go [hunk 13]
```diff
@@ -107,12 +103,12 @@ func (s *Server) Close() {
 // register all server
 func (s *Server) register() error {
 	coreServer := core.NewServer(s.sdk)
-	newCoreServer := newcore.NewServer(s.serviceSDK)
+	coreServiceServer := NewServiceServer(s.sdk)
```
Comment (antho1404) [inline server/grpc/server.go]: could pass only the `s.sdk.Service` to reduce dependencies but it's fine ;)

File: protobuf/api/service.proto (hunk 10 above)
Comment (antho1404) [reply protobuf/api/service.proto]: hum that's annoying. Ultimately the service api (with emitEvent, listenTask) will be removed in favor of the `execution/stream` api and a `core/emit` api. So I'm ok to keep it like that and rename it when we will be at this step.
Maybe if you can just add a TOFIX with a comment to make sure we don't forget about it

File: server/grpc/server.go (hunk 13 above)
Comment (ilgooz) [reply server/grpc/server.go]: 👍 on forum proposal it says that sub sdk packages can use each other so this why chose to use sdk

File: sdk/sdk.go (hunk 11 above)
Comment (ilgooz) [reply sdk/sdk.go]: done

File: server/grpc/api/service.go (hunk 12 above)
Comment (ilgooz) [reply server/grpc/api/service.go]: done

File: protobuf/api/service.proto (hunk 10 above)
Comment (ilgooz) [reply protobuf/api/service.proto]: done

File: sdk/service/service.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+		return err
+	}
+
+	// calculate and apply hash to service.
+	dh := dirhash.New(path)
+	h, err := dh.Sum(nil)
+	if err != nil {
+		return err
+	}
+	srv.Hash = base58.Encode(h)
```
Comment (NicolasMahe) [inline sdk/service/service.go]: Careful here, the service hash should be calculated from the source (already done here) but also from the data of the structure service itself.

File: service/service.go [hunk 15]
```diff
@@ -53,6 +53,9 @@ type Service struct {
 	// a Git host.
 	Repository string `hash:"name:7"`
 
+	// Source is the hash id of service's source code on IPFS.
+	Source string
```
Comment (NicolasMahe) [inline service/service.go]:
```suggestion
	Source string `hash:"-"`
```
Source should be omitted from hash calculation until the cli can generate fully deterministic archive and get the same ifps hash all the time

File: sdk/service/service.go (hunk 14 above)
Comment (NicolasMahe) [reply sdk/service/service.go]: I updated the post https://forum.mesg.com/t/service-compilation-deploy/288 as it didn't explicitly say that the hash calculation changed.

File: service/service.go (hunk 15 above)
Comment (ilgooz) [reply service/service.go]: done

File: sdk/service/service.go (hunk 14 above)
Comment (ilgooz) [reply sdk/service/service.go]: done

**Comment by ilgooz** [comment]
@NicolasMahe let's merge this one, i'll send some fixes in a separate PR.

**Pull request merged**